core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java

```diff
@@ -0,0 +1,118 @@
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.digitalpebble.stormcrawler.protocol.selenium;
+
+import com.digitalpebble.stormcrawler.util.ConfUtils;
+import java.net.URL;
+import java.time.Duration;
+import java.util.List;
+import java.util.Timer;
+import java.util.TimerTask;
+import org.apache.storm.Config;
+import org.openqa.selenium.Dimension;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+public class SeleniumGridImpl extends SeleniumGridProtocol {
+
+    private TimerTask timerTask;
+    private Timer updateQueue;
+    private int noOfWorkers = 0;
+    private String gridAddress;
+    private final DesiredCapabilities capabilities = new DesiredCapabilities();
+
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        noOfWorkers = ConfUtils.getInt(conf, "topology.workers", 2);
+        gridAddress = super.gridAddress;
+        capabilities.setBrowserName(
+                ConfUtils.getString(conf, "selenium.capabilities.browserName", "chrome"));
+        capabilities.setCapability("newSessionWaitTimeout", 600);
+        capabilities.setCapability("browserTimeout", 600);
+        updateQueue = new Timer();
+        updateQueueOfBrowsers();
+    }
+
+    protected void updateQueueOfBrowsers() {
+        timerTask =
+                new TimerTask() {
+                    @Override
+                    public void run() {
+                        try {
+                            List list = getAllNodesList();
+                            LOG.info("Blocking Queue size: " + driversQueue.size());
+                            while (getSessionsCount(list) > 0) {
+                                int size = list.size();
+                                // Check if the queue size is more than the actual
+                                // no of browsers allowed per worker
+                                // means crawler services are idle because all drivers are in queue
+                                if (driversQueue.size() >= (size / noOfWorkers)) {
+                                    SeleniumGridProtocol.Holder holder = driversQueue.take();
+                                    long totalTime = System.currentTimeMillis() - holder.getTime();
+                                    // clear the queue if the total time spent in the queue is more
+                                    // than 4.5 minutes
+                                    // As after 5 mintues the driver would be idle and will throw
+                                    // exception
+                                    // while we try to use that driver for fetching
+                                    if (totalTime > 1000 * 4.5 * 60) {
+                                        driversQueue.clear();
```

Comment:
Why do we clear the full queue? My impression was, that the queue is used to hold `Driver` instances, so we should just remove that specifc driver or did I understand something wrong?

Reply:
Say when the crawler is in a idle mode and has not been running any crawls, the drivers would be sitting  idle and would have lost the session . 

To check every driver session id would be a costly affair. So its better if we clear the queue and again add it from the grid which are active and available in the grid.

Reply:
Here two sets of list are getting  handled one the grid handles and the other queue we handle per worker. So its better to clear out the queue and try adding it again from the available nodes from the grid.